src/cache/memcache.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 Mozilla Foundation
+// Copyright 2017 David Michael Barr <b@rr-dav.id.au>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use cache::{
+    Cache,
+    CacheRead,
+    CacheWrite,
+    Storage,
+};
+use errors::*;
+use futures::Future;
+use futures_cpupool::CpuPool;
+use memcached::client::Client;
+use memcached::proto::Error as ProtoError;
+use memcached::proto::{
+    Operation,
+    ProtoType
+};
+use std::io::Cursor;
+use std::time::{
+    Duration,
+    Instant,
+};
+
+#[derive(Clone)]
+pub struct MemCache {
+    url: String,
+    pool: CpuPool,
+}
+
+impl MemCache {
+    pub fn new(url: &str, pool: &CpuPool) -> Result<MemCache> {
+        Ok(MemCache {
+            url: url.to_owned(),
+            pool: pool.clone(),
+        })
+    }
+
+    //TODO: split and parse server list
+    fn connect(&self) -> Result<Client> {
+        Client::connect(&[(&self.url, 1)], ProtoType::Binary)
+            .map_err(|e| e.into())
+    }
+}
+
+impl Storage for MemCache {
+    fn get(&self, key: &str) -> SFuture<Cache> {
+        let key = key.to_owned();
+        let me = self.clone();
+        self.pool.spawn_fn(move || {
+            let mut c = me.connect().unwrap();
```

Comment:
I had the same thoughts with the Redis cache and [tried](https://github.com/flxo/sccache/commit/25118045c477fe76bfdcd9b0f6649644df543e64) out  [r2d2](https://github.com/sfackler/r2d2). 

We have had problems with the thousands of opened/closed connections when running sccache within docker. It's not yet clear in detail why - still investigating. That's why I added the pooling in our internal test branch.